### DISCUSSION TITLE
SQLAlchemy incorrectly defines the path to the class in relationship

### DISCUSSION THREAD

**Discussion by klaydev** [sqlalchemy/sqlalchemy#10334] [Usage Questions]
Hi, i have this code (just example):
```py
# a_model.py
class A(DeclarativeBase): # DeclarativeBase - custom instance of DeclarativeBase (func declarative_base)
    . . .
    b_id = mapped_column(Integer, ForeignKey("b.id"))
    b = relationship("src.models.b_model.B", back_populates="a_list")
    . . .
 ```
 ```py
 # b_model.py
 class B(DeclarativeBase): # DeclarativeBase - custom instance of DeclarativeBase (func declarative_base)
     . . .
     a_list = relationship("src.models.a_model.A", back_populates="b")
     . . .
```
And i get this error (real error):
```py
sqlalchemy.exc.InvalidRequestError: When initializing mapper Mapper[ShopModel(shops)], expression 'src.models.server.ServerModel' failed to locate a name ("Module 'models' has no mapped classes registered under the name 'server'"). If this is a class name, consider adding this relationship() to the <class 'src.models.shop.ShopModel'> class after both dependent classes have been defined.
```
As I understand it, SQLAlchemy believes that models is a module, although it is a directory (diagram below)
```
src/
 |- models/
      |- a_model.py
           |- A
      |- b_model.py
           |- B
 ```

**Comment by zzzeek** [comment]
in python, when you refer to a name like "a.b.c.SomeClass", that means that `a` and `b` are both packages, `c` is a module, and `SomeClass` is a class.    Packages all must have an `__init__.py` present for them to be locatable as packages.   

as far as SQLAlchemy's error, it might be worded incorrectly in this case, but the main thing is that it is not finding a symbol named ".server" in your ".models" package, and you are specifying a path as `'src.models.server.ServerModel'` somewhere.

**Reply by klaydev** [reply]
Hm. I use this form of treatment, because when importing directly (from src.models.server import Server Model)  I will get an error about circular import. It is very strange that the behavior is like this, because I have specified the full path to the class

**Reply by zzzeek** [reply]
you can set up SQLAlchemy models in each module without having to import the other module.  However, when you are about to first **use** the models, you must `import` all dependent modules first so that all the dependent mappings can be picked up.
